src/main/java/com/sonymobile/jenkins/plugins/mq/mqnotifier/MQConnection.java

```diff
@@ -207,20 +240,115 @@ public void publish(JSONObject json) {
      * Sends messages from the message queue.
      */
     private void sendMessages() {
+        Channel channel = null;
+
         while (true) {
             try {
-                MessageData messageData = (MessageData)messageQueue.poll(SENDMESSAGE_TIMEOUT,
+                if (channel == null || !channel.isOpen()) {
+                    channel = createChannel();
+                    channel.confirmSelect();
+                    addMessageConfirmListener(channel);
+                    setShouldAddToQueue(true);
+                }
+                MessageData messageData = (MessageData) messageQueue.poll(SENDMESSAGE_TIMEOUT,
                                                                          TimeUnit.MILLISECONDS);
                 if (messageData != null) {
-                    getInstance().send(messageData.getExchange(), messageData.getRoutingKey(),
-                            messageData.getProps(), messageData.getBody());
+                    validateExchange(channel, messageData.getExchange());
+                    getInstance().sendOnChannel(messageData, channel);
                 }
             } catch (InterruptedException ie) {
                 LOGGER.info("sendMessages() poll() was interrupted: ", ie);
+            } catch (IOException ioe) {
+                LOGGER.error("error validating channel: ", ioe);
+            } catch (ChannelCreationException | MessageDeliveryException transientException) {
+                LOGGER.error(transientException.getMessage(), transientException.getCause());
+                setShouldAddToQueue(false);
+                try {
+                    Thread.sleep(CONNECTION_WAIT);
+                } catch (InterruptedException ie) {
+                    LOGGER.error("Thread.sleep() was interrupted", ie);
+                }
+            }
+        }
+    }
+
+
+    /**
+     * Enable or disable adding messages to the queue.
+     */
+    private synchronized void setShouldAddToQueue(boolean enabled) {
+        this.shouldAddToQueue = enabled;
+        notifyAll();
+    }
+
+    /**
+     * Wait until it's ok to add messages to the queue again
+     */
+    public synchronized void waitForQueue() {
+        while (!this.shouldAddToQueue) {
+            try {
+                wait();
+            } catch (InterruptedException e)  {
+                Thread.currentThread().interrupt();
+            }
```

Comment:
Will this work? Wont setting the interrupt just cause the wait() to throw another interruption?

Reply:
Yes. Used the sample from:
https://www.baeldung.com/java-wait-notify in section 5

Reply:
Yes, it seems logical the Thread.currentThread().interrupt() shouldn't be there, which lowers my trust for Baeldung...

The code should probably be:

synchronized(a) {
    try{
        while(true) {
            a.wait()
        }
    }catch(InterruptedException...){
        ...
    }
}

as in https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html.

Although, this piece of code was removed entirely in the recently submitted changes...